src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+    Node_List* stmt = _stmts.at(idx);
+    for (int i = stmt->size() - 1; i >= 0; i--) {
+      Node* node = stmt->at(i);
+      int opc = node->Opcode();
+      BasicType bt = elem_bt(node);
+      int vlen = Matcher::max_vector_size(bt);
+      if (vlen == 0) {
+        // Bail out if vector cannot hold such elements
+        return false;
+      }
+      // We check special convert and min/max ops first and then general ops
+      if (VectorNode::is_convert_opcode(opc)) {
+        Node* in = node->in(1);
+        BasicType in_bt = is_loop_iv_or_incr(in) ? T_INT : elem_bt(in);
+        if (in_bt == T_ILLEGAL || !same_element_size(in_bt, bt) ||
+            !VectorCastNode::implemented(opc, vlen, in_bt, bt)) {
+          trace_msg(node, "Found unimplemented vector cast node");
+          return false;
+        }
+      } else if (VectorNode::is_minmax_opcode(opc) && is_subword_type(bt)) {
+        // Java API for Math.min/max operations supports only int, long, float
+        // and double types. Bail out for subword min/max operations.
+        return false;
+      } else {
+        int vopc = 0;
+        if (node->is_Mem()) {
+          vopc = node->is_Store() ? Op_StoreVectorMasked : Op_LoadVectorMasked;
+        } else {
+          vopc = VectorNode::opcode(opc, bt);
+        }
+        if (vopc == 0 ||
+            !Matcher::match_rule_supported_vector_masked(vopc, vlen, bt)) {
+          trace_msg(node, "Vector replacement node is not implemented");
+          return false;
+        }
+      }
+    }
+  }
+  return true;
+}
+
+// Find unhandled out-of-loop use of loop body nodes and untracked loop body
+// nodes to bail out for complex loops
+bool VectorMaskedLoop::analyze_loop_body_nodes() {
+  VectorSet tracked(_arena);
+  int n_nodes = _body_nodes.length();
+  // 1) Track all vectorization candidates and loop iv phi nodes
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* node = _body_nodes.at(idx);
+    if (has_valid_elem_bt(node) || is_loop_iv(node)) {
+      tracked.set(idx);
+    }
+  }
+  // 2) Track memory address computing nodes in SWPointer node stacks
+  for (int ptridx = 0; ptridx < _swptrs.length(); ptridx++) {
+    Node_Stack* nstack = _swptrs.at(ptridx)->node_stack();
+    while (nstack->is_nonempty()) {
+      Node* node = nstack->node();
+      if (in_body(node)) {
+        tracked.set(rpo_idx(node));
+      }
+      nstack->pop();
+    }
+  }
+  // 3) Up to this point, all tracked nodes shouldn't have out-of-loop users
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* node = _body_nodes.at(idx);
+    if ((node->is_Mem() && node->as_Mem()->is_Store())) {
+      // Only store nodes are exceptions
+      continue;
+    }
+    if (tracked.test(idx)) {
+      for (DUIterator_Fast imax, i = node->fast_outs(imax); i < imax; i++) {
+        Node* out = node->fast_out(i);
+        if (!in_body(out)) {
+          trace_msg(node, "Node has out-of-loop user found");
+          return false;
```

Comment:
Can this be handled in the future with a extract node? I guess you would have to extract it from a variable element, as the last iteration is not always the same.

Reply:
Probably, but I haven't tried it so far. Will do it in the future.